// appium-espresso-driver/build/test/functional/desired.js
'use strict';

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _gpsDemoApp = require('gps-demo-app');

var _gpsDemoApp2 = _interopRequireDefault(_gpsDemoApp);

var apidemosApp = require.resolve('android-apidemos');

var GENERIC_CAPS = {
  androidInstallTimeout: 90000,
  deviceName: 'Android',
  platformName: 'Android',
  forceEspressoRebuild: true,
  adbExecTimeout: process.env.CI ? 120000 : 20000,
  espressoServerLaunchTimeout: process.env.CI ? 120000 : 30000,
  printPageSourceOnFindFailure: true
};

var APIDEMO_CAPS = _Object$assign({}, GENERIC_CAPS, {
  app: apidemosApp
});

var GPS_CAPS = _Object$assign({}, GENERIC_CAPS, {
  app: _gpsDemoApp2['default']
});

var REACT_NATIVE_CAPS = _Object$assign({}, GENERIC_CAPS, {
  app: _path2['default'].resolve(__dirname, '..', '..', 'assets', 'ReactNativeApp.apk')
});

exports.GENERIC_CAPS = GENERIC_CAPS;
exports.APIDEMO_CAPS = APIDEMO_CAPS;
exports.GPS_CAPS = GPS_CAPS;
exports.REACT_NATIVE_CAPS = REACT_NATIVE_CAPS;
